db/types/operations/cast.py

```diff
@@ -962,6 +973,123 @@ def _get_text_uri_type_body_map():
     return {type_: _get_text_uri_type_body_map() for type_ in source_types}
 
 
+def _get_numeric_type_body_map():
+    """
+    Get SQL strings that create various functions for casting different
+    types to numeric.
+    We allow casting any textual type to locale-agnostic numeric.
+    """
+    default_behavior_source_types = NUMBER_TYPES | frozenset([money.DB_TYPE])
+    text_source_types = TEXT_TYPES
+
+    type_body_map = _get_default_type_body_map(
+        default_behavior_source_types, NUMERIC
+    )
+    type_body_map.update(
+        {
+            text_type: _get_text_to_numeric_cast()
+            for text_type in text_source_types
+        }
+    )
+    type_body_map.update({BOOLEAN: _get_boolean_to_number_cast(NUMERIC)})
+    return type_body_map
+
+
+def _get_text_to_numeric_cast():
+    numeric_array_function = base.get_qualified_name(NUMERIC_ARR_FUNC_NAME)
+    cast_exception_str = (
+        f"RAISE EXCEPTION '% cannot be cast to {NUMERIC}', $1;"
+    )
+    return rf"""
+    DECLARE decimal_point {TEXT};
+    DECLARE is_negative {BOOLEAN};
+    DECLARE numeric_arr {TEXT}[];
+    DECLARE numeric {TEXT};
+    BEGIN
+        SELECT {numeric_array_function}($1::{TEXT}) INTO numeric_arr;
+        IF numeric_arr IS NULL THEN
+            {cast_exception_str}
+        END IF;
+
+        SELECT numeric_arr[1] INTO numeric;
+        SELECT ltrim(to_char(1, 'D'), ' ') INTO decimal_point;
+        SELECT $1::text ~ '^-.*$' INTO is_negative;
+
+        IF numeric_arr[2] IS NOT NULL THEN
+            SELECT regexp_replace(numeric, numeric_arr[2], '', 'gq') INTO numeric;
+        END IF;
+        IF numeric_arr[3] IS NOT NULL THEN
+            SELECT regexp_replace(numeric, numeric_arr[3], decimal_point, 'q') INTO numeric;
+        END IF;
+        IF is_negative THEN
+            RETURN ('-' || numeric)::{NUMERIC};
+        END IF;
+        RETURN numeric::{NUMERIC};
+    END;
+    """
+
+
+def _build_numeric_array_function():
+    """
+    The main reason for this function to be separate is for testing. This
+    does have some performance impact; we should consider inlining later.
+    """
+    qualified_function_name = base.get_qualified_name(NUMERIC_ARR_FUNC_NAME)
+
+    no_separator = r"[0-9]+(?:([.,])[0-9]+)?"
```

Comment:
Yeah, you're right. I guess I was too focused on covering all the possible cases that I also forgot to consider false positives (the cases that represent what they are not, in fact).
And still, what about examples 123,456 and 123.456 that you provided at the beginning? I guess it won't be ok just to skip them because they definitely represent numbers. Although, I understand yout concern about them not being handled correctly.

Reply:
@mathemancer guess you just missed this question. May I get your opinion on what we will do with numbers like `123,456` or `123.456`? And I think that such format `-123'456.7` (`de-CH`) is also not handled in the `build_mathesar_money_array` function. But should it?

Reply:
Ah, sorry. I really wish GH would highlight new comments somehow.

For the moment, we should not cast `123,456` or `123.456`. While those _are_ numbers, casting in the context of inference has the potential to irreversibly corrupt data as described above. We'll eventually be getting more and more sophisticated about these things, and should be able to add those later. If the user wants to cast a column with such values, they'll have to do some manual work removing commas or adding `.0` or `,0` to each value.

As for `de-CH`, you're right; good catch! Please add that to this PR, and raise a bug for the money type.